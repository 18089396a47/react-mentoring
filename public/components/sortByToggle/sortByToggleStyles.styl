.sort-by-toggle
  display: flex;
  align-items: center;

  &__option
    margin-left: 15px;
    overflow: hidden;

    input[type=radio]
      display: none;

  &__option-title
    cursor: pointer;
    transition: color 200ms;

    &:hover
      color: #363636;

    :checked + &
      color: #f45162;

      &:hover
        color: #e03d4e;
