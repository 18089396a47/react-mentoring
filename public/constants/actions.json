{
  "searchStart": "SEARCH_START",
  "searchEnd": "SEARCH_END",
  "changeSearchInput": "CHANGE_SEARCH_INPUT",
  "updateSearchResults": "UPDATE_SEARCH_RESULTS"
}
