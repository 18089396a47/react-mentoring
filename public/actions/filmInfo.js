import { call, put, all, takeLatest } from 'redux-saga/effects';
import * as types from '../constants/actions';
import theMovieDb from 'themoviedb-javascript-library';
import * as commons from '../constants/commons';
import promiseWrapper from '../helpers/promiseWrapper';

export const searchFilmEnd = () => ({
    type: types.searchFilmEnd
});

export const searchFilmStart = (id, searchType) => ({
    type: types.searchFilmStart,
    id,
    searchType
});

function* fetchFilm({ id, searchType }) {
    const searchMethod = searchType === commons.searchTypeMovie ?
        theMovieDb.movies.getById : theMovieDb.tv.getById;

    try {
        const response = yield call(() => promiseWrapper(searchMethod, {
            id
        }));

        yield put(searchFilmEnd());
        yield put(updateFilmInfo(JSON.parse(response)))
    } catch (e) {
        yield put(searchFilmEnd());
    }
}

function* watchFetchFilm() {
    yield takeLatest(types.searchFilmStart, fetchFilm);
}

export function* filmInfoSaga() {
    yield all([
        watchFetchFilm()
    ]);
}

export const updateFilmInfo = (data) => ({
    type: types.updateFilmInfo,
    data
});
