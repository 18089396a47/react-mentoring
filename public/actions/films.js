import * as types from '../constants/actions';

export const updateSearchResults = (data) => ({
    type: types.updateSearchResults,
    data: JSON.parse(data)
});
