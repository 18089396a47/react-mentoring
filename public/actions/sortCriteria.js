import * as types from '../constants/actions';

export const changeSortCriteria = (sortCriteria) => ({
    type: types.changeSortCriteria,
    sortCriteria
});
