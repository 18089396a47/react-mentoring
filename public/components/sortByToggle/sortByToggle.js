import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import './sortByToggleStyles';
import { changeSortCriteria } from '../../actions/';
import * as commons from '../../constants/commons';

class SortByToggle extends Component {
    changeCriteria(criteria) {
        this.props.dispatch(changeSortCriteria(criteria));
    }

    render() {
        return (
            <div className="sort-by-toggle">
                <span className="sort-by-toggle__title">Sort by</span>
                <label className="sort-by-toggle__option">
                    <input
                        type="radio"
                        name="sort-by-toggle"
                        checked={this.props.sortCriteria === commons.sortByDate}
                        onChange={this.changeCriteria.bind(this, commons.sortByDate)}
                    />
                    <span className="sort-by-toggle__option-title">release date</span>
                </label>
                <label className="sort-by-toggle__option">
                    <input
                        type="radio"
                        name="sort-by-toggle"
                        checked={this.props.sortCriteria === commons.sortByRating}
                        onChange={this.changeCriteria.bind(this, commons.sortByRating)}
                    />
                    <span className="sort-by-toggle__option-title">rating</span>
                </label>
            </div>
        );
    }
}

SortByToggle.propTypes = {
    sortCriteria: PropTypes.string.isRequired,
    dispatch: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
    return {
        sortCriteria: state.sortCriteria
    };
};

export default connect(mapStateToProps)(SortByToggle);
