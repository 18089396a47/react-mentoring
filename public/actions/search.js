import { call, put, all, takeLatest } from 'redux-saga/effects';
import * as types from '../constants/actions';
import theMovieDb from 'themoviedb-javascript-library';
import { updateSearchResults } from './films';
import * as commons from '../constants/commons';
import promiseWrapper from '../helpers/promiseWrapper';

export const changeSearchInput = (inputValue) => ({
    type: types.changeSearchInput,
    inputValue
});

export const searchQueryEnd = () => ({
    type: types.searchQueryEnd
});

export const searchQueryStart = (inputValue, searchType) => ({
    type: types.searchQueryStart,
    inputValue,
    searchType
});

function* fetchQuery({ inputValue, searchType }) {
    const searchMethod = searchType === commons.searchTypeMovie ?
        theMovieDb.search.getMovie : theMovieDb.search.getTv;

    try {
        const response = yield call(() => promiseWrapper(searchMethod, {
            query: inputValue
        }));

        yield put(searchQueryEnd());
        yield put(updateSearchResults(JSON.parse(response)))
    } catch (e) {
        yield put(searchQueryEnd());
    }
};

function* watchFetchQuery() {
    yield takeLatest(types.searchQueryStart, fetchQuery);
}

export const searchSimilarFilmEnd = () => ({
    type: types.searchSimilarFilmEnd
});

export const searchSimilarFilmStart = (id) => ({
    type: types.searchSimilarFilmStart,
    id
});

function* fetchSimilarFilms({ id }) {
    try {
        const response = yield call(() => promiseWrapper(theMovieDb.movies.getSimilarMovies, {
            id
        }));

        yield put(searchSimilarFilmEnd());
        yield put(updateSearchResults(JSON.parse(response)))
    } catch (e) {
        yield put(searchSimilarFilmEnd());
    }
};

function* watchFetchSimilarFilms() {
    yield takeLatest(types.searchSimilarFilmStart, fetchSimilarFilms);
}

export function* searchSaga() {
    yield all([
        watchFetchSimilarFilms(),
        watchFetchQuery()
    ]);
}

export const changeSearchType = (searchType) => ({
    type: types.changeSearchType,
    searchType
});
