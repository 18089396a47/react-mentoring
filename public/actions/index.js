import { all } from 'redux-saga/effects';

import { filmInfoSaga } from './filmInfo';
import { searchSaga } from './search';

export default function* rootSaga() {
    yield all([
        filmInfoSaga(),
        searchSaga()
    ]);
}

export * from './search';
export * from './sortCriteria';
export * from './filmInfo';
export * from './films';
