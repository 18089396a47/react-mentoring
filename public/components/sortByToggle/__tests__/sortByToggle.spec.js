import React from 'react';
import renderer from 'react-test-renderer';
import { SortByToggle } from '../sortByToggle';
import { changeSortCriteria } from '../../../actions/';

const mockDigest = jest.fn();

jest.mock('../../../actions/index', () => ({
    changeSortCriteria: jest.fn()
}));

let props;

describe('SortByToggle component', () => {
    beforeEach(() => {
        props = {
            dispatch: mockDigest,
            sortCriteria: 'SORT_BY_DATE'
        };
    });

    it('SortByToggle should be shown correctly', () => {
        const tree = renderer.create(
            <SortByToggle {...props} />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('SortByToggle should call \'changeSortCriteria\' redux action on radio change', () => {
        const tree = shallow(<SortByToggle {...props} />);
        const radio = tree.find('.sort-by-toggle__option input').last();

        radio.simulate('change', {
            target: {
                checked: true
            }
        });
        expect(mockDigest).toHaveBeenCalled();
        expect(changeSortCriteria).toHaveBeenCalledWith('SORT_BY_RATING');
    });
});
