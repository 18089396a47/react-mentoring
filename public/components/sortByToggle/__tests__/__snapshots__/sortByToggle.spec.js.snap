// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortByToggle component SortByToggle should be shown correctly 1`] = `
<div
  className="sort-by-toggle"
>
  <span
    className="sort-by-toggle__title"
  >
    Sort by
  </span>
  <label
    className="sort-by-toggle__option"
  >
    <input
      checked={true}
      name="sort-by-toggle"
      onChange={[Function]}
      type="radio"
    />
    <span
      className="sort-by-toggle__option-title"
    >
      release date
    </span>
  </label>
  <label
    className="sort-by-toggle__option"
  >
    <input
      checked={false}
      name="sort-by-toggle"
      onChange={[Function]}
      type="radio"
    />
    <span
      className="sort-by-toggle__option-title"
    >
      rating
    </span>
  </label>
</div>
`;
